File: code/python/hash/linear_probe_hash.py
class LinearProbeHash:
    """A hash symbol table using the linear probing."""

    def __init__(self, cap=4):
        self._n = 0
        self._m = cap
        self._keys = [None] * cap
        self._values = [None] * cap

    def size(self):
        return self._n

    def is_empty(self):
        return self._n == 0

    def _hash(self, k):
        return hash(k) % self._m

    def __setitem__(self, key, value):
        if key is None:
            raise ValueError
        if value is None:
            del self[key]
            return
        if self._n >= self._m // 2:
            self._resize(self._m * 2)
        i = self._hash(key)
        while self._keys[i] is not None:
            if key == self._keys[i]:
                self._values[i] = value
                return  # update
            i = (i + 1) % self._m
        # insert a new key-value pair
        self._keys[i] = key
        self._values[i] = value
        self._n += 1

    def __getitem__(self, key):
        if key is None:
            raise ValueError
        i = self._hash(key)
        while self._keys[i] is not None:
            if key == self._keys[i]:
                return self._values[i]
            i = (i + 1) % self._m
        return None

    def contains(self, key):
        if key is None:
            raise ValueError
        return self[key] is not None

    def __delitem__(self, key):
        if key is None:
            raise ValueError
        if self[key] is None:
            return  # no such key
        # find the position
        i = self._hash(key)
        while self._keys[i] != key:
            i = (i + 1) % self._m
        # delete key and associated value
        self._keys[i] = None
        self._values[i] = None
        # rehash all keys in the same cluster
        i = (i + 1) % self._m
        while self._keys[i] is not None:
            key_to_rehash = self._keys[i]
            value_to_rehash = self._values[i]
            self._keys[i] = None
            self._values[i] = None
            self._n -= 1
            self[key_to_rehash] = value_to_rehash
            i = (i + 1) % self._m
        self._n -= 1
        if 0 < self._n <= self._m // 8:
            self._resize(self._m / 2)

    def _resize(self, cap):
        t = LinearProbeHash(cap)
        for i in range(0, self._m):
            if self._keys[i] is not None:
                t[self._keys[i]] = self._values[i]
        self._m = cap
        self._keys = t._keys
        self._values = t._values

File: code/java/hash/src/main/java/org/swufe/datastructures/LinearProbeHash.java
package org.swufe.datastructures;

public class LinearProbeHash<Key, Value> {

    private static final int INIT_CAPACITY = 4; // 2's power

    private Key[] keys;
    private Value[] values;

    private int n; // number of key-value pairs
    private int m; // size of hash table

    @SuppressWarnings("unchecked")
    public LinearProbeHash(int cap) {
        keys = (Key[]) new Object[cap];
        values = (Value[]) new Object[cap];
        m = cap;
        n = 0;
    }

    public LinearProbeHash() {
        this(INIT_CAPACITY);
    }

    public int size() {
        return n;
    }

    public boolean isEmpty() {
        return n == 0;
    }
    private int hash(Object x) {
        int h = x.hashCode();
        h ^= (h >>> 16);
        return h & (m - 1);
    }

    public void put(Key k, Value v) {
        if (k == null) {
            throw new IllegalArgumentException();
        }
        if (v == null) {
            delete(k);
            return;
        }
        if (n >= m/2) {
            resize(2 * m);
        }
        int i;
        for (i = hash(k); keys[i] != null; i = (i + 1) % m) {
            if (keys[i].equals(k)) {
                values[i] = v;
                return;
            }
        }
        keys[i] = k;
        values[i] = v;
        n++;
    }

    public Value get(Key k) {
        for (int i = hash(k); keys[i] != null; i = (i + 1) % m) {
            if (keys[i].equals(k)) {
                return values[i];
            }
        }
        return null;
    }

    public boolean contains(Key k) {
        if (k == null) throw new IllegalArgumentException();
        return get(k) != null;
    }

    public void delete(Key k) {
        if (!contains(k)) return;
        // find the position of `k`
        int i = hash(k);
        while (!k.equals(keys[i])) {
            i = (i + 1) % m;
        }
        // delete key and associated value
        keys[i] = null;
        values[i] = null;
        // rehash all keys in the same cluster
        i = (i + 1) % m;
        while (keys[i] != null) {
            Key keyToRehash = keys[i];
            Value valueToRehash = values[i];
            keys[i] = null;
            values[i] = null;
            n--;
            put(keyToRehash, valueToRehash);
            i = (i + 1) % m;
        }
        n--;
        if (n > 0 && n <= m / 8) resize(m / 2);
    }
    private void resize(int cap) {
        LinearProbeHash<Key, Value> t = new LinearProbeHash<>(cap);
        for (int i = 0; i < m; i++) {
            if (keys[i] != null) {
                t.put(keys[i], values[i]);
            }
        }
        keys = t.keys;
        values = t.values;
        m = t.m;
    }
}
